src/mlpack/tests/main_tests/decision_tree_test.cpp

```diff
@@ -88,6 +90,49 @@ BOOST_AUTO_TEST_CASE(DecisionTreeOutputDimensionTest)
   BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("probabilities").n_rows, 3);
 }
 
+/**
+ * Check that number of output points and number
+ * of input points are equal for categorical dataset.
+ */
+BOOST_AUTO_TEST_CASE(DecisionTreeCategoricalOutputDimensionTest)
+{
+  arma::mat inputData;
+  DatasetInfo info;
+  if (!data::Load("t1.arff", inputData, info))
+    BOOST_FAIL("Cannot load train dataset t1.arff!");
+
+  // Initialize an all-ones weight matrix.
+  arma::mat weights(1, inputData.n_cols, arma::fill::ones);
+
+  arma::mat testData;
+  if (!data::Load("t1_test.arff", testData, info))
+    BOOST_FAIL("Cannot load test dataset t1_test.arff!");
+
+  size_t testSize = testData.n_cols;
+
+  // Input training data.
+  SetInputParam("training", std::move(std::make_tuple(info, inputData)));
+  // Get labels from last column of training data.
+  SetInputParam("weights", std::move(weights));
+
+  // Input test data.
+  SetInputParam("test", std::move(std::make_tuple(info, testData)));
+
+  mlpackMain();
+
+  // Check that number of output points are equal to number of input points.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Row<size_t>>("predictions").n_cols,
+      testSize);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("probabilities").n_cols,
+      testSize);
+
+  // Check number of output rows equals number of classes in case of
+  // probabilities and 1 for predictions.
+  BOOST_REQUIRE_EQUAL(
+      CLI::GetParam<arma::Row<size_t>>("predictions").n_rows, 1);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("probabilities").n_rows, 6);
+}
```

Comment:
Can I ask you to add one more test please?  I want to make sure that the DatasetInfo mappings are properly loaded for saved models.

So the test could build a model on some categorical data in one run and store predictions on test data, and then also store the `output_model`.  Then the stored model could be given as `input_model` with a test set, and we should ensure that the predictions and class probabilities are identical with both runs (you can use the `CheckMatrices()` function for that).

Let me know what you think or if I can clarify what I mean any further.

Reply:
I did try this manually as per your description in the issue itself and it is working as desired. I agree, a proper test will be good though. I'll work on it. :+1: